### DISCUSSION TITLE
Is Talos suitable for use without Kubernetes

### DISCUSSION THREAD

**Discussion by kurtismullins** [siderolabs/talos#10008] [Q&A]
I'm interested in provisioning machines with an immutable OS since I only need to run a few processes on each machine and the machines (aside from an attached volume/drive) can be ephemeral. I don't need a full Kubernetes environment. I also don't want to use Ansible or other configuration management utilities because the machine's state can sway from the intended config over time; it's much easier to just re-provision your machine with a new init script.

Is anyone using Talos for this purpose? Are there any downsides specific to Talos, especially compared to other immutable OS distributions, for doing this kind of thing?

Thanks!

**Comment by smira** [comment]
This is something Talos Linux can certainly do, but there are some missing pieces to make it fully Kubernetes-less.

I haven't tried that, but you could do a single controlplane node, disable kubelet node registration, skip controller-manager & scheduler (api-server will still be started), and use `.machine.pods` to launch static pods.

This way nothing can be scheduled on the machine using Kubernetes APIs, but you can manage your workloads modifying `.machine.pods`.

**Comment by rothgar** [comment]
Thanks for the suggestion Kurt. It's absolutely something that has been on our minds and an eventual long term goal of having an immutable, API driven, single purpose Linux distribution. We strongly believe that making Linux do 1 thing makes them more stable and maintainable long term.

But anything we build on top of Talos would be a bit of a hack. Talos will always stay a single-purpose Kubernetes distribution. We would create different single-purpose distributions for other use cases so we don't overly complicate the machine API and the distros can remain focused.

Could you let us know how you'd like to distribute your software to the nodes (e.g. containers, binaries)? Do you expect machines to be highly available or part of a cluster, not Kubernetes? Do you want centralized OS management (eg Omni) or is this for individual use that you would manage 1-by-1?

Right now we don't have any plans to build this but we're gathering data about use cases to see what we can do in the future.

**Reply by kurtismullins** [reply]
> Talos will always stay a single-purpose Kubernetes distribution

Thanks for letting me know!

> Could you let us know how you'd like to distribute your software to the nodes (e.g. containers, binaries)?

Primarily, containers. However, some binaries make sense to run outside of a container (think node-exporter, cadvisor, etc..) just due to the complex configuration required to grant access to everything they need.

> Do you expect machines to be highly available or part of a cluster, not Kubernetes?

Both! In my current employer's case, we do not use Kubernetes because the traditional "deployment" model doesn't work for our software; it handles replication, internal routing, etc... itself. We would affectively just be running N replicas and using Kubernetes features to limit it to a single process per machine and attempting to limit Disk IO to the localhost.  A lot of the functionality would be replicated. This can be a pain to properly configure and requires everyone on-call to understand maintaining Kubernetes systems on bare metal. 

For personal purposes, I could see using it as a lightweight, immutable server OS with a nice API. 

> Do you want centralized OS management (eg Omni)

I have never used Omni but my preference would be more agnostic tooling such as Terraform and Ansible. Preferably just being able to configure everything initially with cloud-init/ignition/etc... would be a wonderful start. 

> is this for individual use that you would manage 1-by-1

Maybe a nice-to-have but definitely not my business case.

**Reply by divStar** [reply]
I know this is an older issue, but my use case is simple: I would like the most clean Linux distro I ever saw (Talos) to be able to just run Docker so I can deploy containers in a far simpler way than Kubernetes (because at home I don't need the Kubernetes overhead everywhere).

I haven't found a Linus distribution, that is so nice to work with like Talos (yes, I've been using many different ones over 2 decades, but I just love the fact, that only the most basic tools are running on Talos - and then Kubernetes). If I could have that with Docker (or postman if need be) instead, I'd probably consider to have found the best base for my containers.

Seriously: it's so easy to configure the OS part of Talos it makes me smile every time.

Regardless of what you guys over at SideroLabs decide: thanks for making Talos! I hope it'll over just Docker one day - and the great benefits of SystemExtensions etc.

**Reply by frezbo** [reply]
@rothgar

**Reply by fullmetal-fred** [reply]
We've had this discussion before, but I just wanted to chime in that for my own use case, just enabling Talos nodes in stand-alone mode (no k8s) with static pod definitions would be plenty. I'd include container defs in the machine config. I'm sure compose would be a nice addition to that. One config file for everything would be killer. And if you wanted to go further, just drop the portainer agent on the static config and you have a standalone docker (containerd) node. Ive gotten very comfortable with k8s now, but for me these use cases are purely for IoT nodes that are set and forget.

**Reply by rothgar** [reply]
Talos is always intended to be used with Kubernetes, but we totally understand the API management makes it the best Linux server distro! We are making progress to make Talos more modular. This includes things like making [network configuration independent from machine config](https://github.com/siderolabs/talos/issues/10925) and eventually we'll make it possible to run external, dedicate etcd nodes.

These changes will help with scaling and configuration in a variety of ways.

Eventually, we'll need a way to make it possible to run Talos _only_ as an etcd cluster and that would require the ability to not run specific kubernetes components.

You can see where this might lead in the future to make it possible to have a machine config with
```
machine:
    type: none
```
Which only runs the the Talos API server and containerd.

We're not there yet, but for us there's no going back to traditional Linux management.

**Reply by polRk** [reply]
I'm really looking forward to it too! You did well!

**Comment by fullmetal-fred** [comment]
@rothgar 

I actually created a discussion on a similar topic a while back: https://github.com/siderolabs/talos/discussions/8343

FWIW: just enabling static pods on the kubelet via machine config & Omni without requiring full cluster formation would be massive for my use case which is more light IoT nodes with a migration bath to full K8s. I think in some ways that would allow Talos to "stay k8s", and with Omni it would allow you to take business from the Balena's of the world who do this on their own Docker distribution. 

Would love to discuss this!
